ostd/src/mm/dma/dma_stream.rs

```diff
@@ -1,212 +1,349 @@
+/// prevents users from reading and writing to [`DmaStream`] unexpectedly.
+pub trait DmaDirection: 'static + Debug + private::Sealed {
+    /// Whether the CPU can read data from the device.
+    const CAN_READ_FROM_DEVICE: bool;
+    /// Whether the CPU can write data to the device.
+    const CAN_WRITE_TO_DEVICE: bool;
+}
+
+mod private {
+    /// To avoid users implement `DmaDirection` and triggers unreachable code in
+    /// functions like [`crate::arch::mm::sync_dma_range`], or bypasses checks
+    /// in [`crate::mm::io_util::HasVmReaderWriter`].
+    pub trait Sealed {}
+}
+
+/// Data flows to the device.
+///
+/// From the perspective of the kernel, this memory region is writable.
+#[derive(Debug)]
+pub enum ToDevice {}
+
+impl private::Sealed for ToDevice {}
+impl DmaDirection for ToDevice {
+    const CAN_READ_FROM_DEVICE: bool = false;
+    const CAN_WRITE_TO_DEVICE: bool = true;
+}
+
+/// Data flows from the device.
+///
+/// From the perspective of the kernel, this memory region is read-only.
+#[derive(Debug)]
+pub enum FromDevice {}
+
+impl private::Sealed for FromDevice {}
+impl DmaDirection for FromDevice {
+    const CAN_READ_FROM_DEVICE: bool = true;
+    const CAN_WRITE_TO_DEVICE: bool = false;
+}
+
+/// Data flows both from and to the device.
+#[derive(Debug)]
+pub enum Bidirectional {}
+
+impl private::Sealed for Bidirectional {}
+impl DmaDirection for Bidirectional {
+    const CAN_READ_FROM_DEVICE: bool = true;
+    const CAN_WRITE_TO_DEVICE: bool = true;
+}
+
+/// A DMA memory object with streaming access.
+///
+/// The kernel must synchronize the data by [`sync`] when interacting with the
+/// device.
 ///
-/// Users must synchronize data before reading or after writing to ensure
-/// consistency.
+/// [`sync`]: DmaStream::sync
 #[derive(Debug)]
-pub struct DmaStream {
-    segment: USegment,
-    start_daddr: Daddr,
+pub struct DmaStream<D: DmaDirection = Bidirectional> {
+    inner: ManuallyDrop<StreamInner>,
+    map_daddr: Option<Daddr>,
     is_cache_coherent: bool,
-    direction: DmaDirection,
+    _phantom: PhantomData<D>,
 }
 
-/// `DmaDirection` limits the data flow direction of [`DmaStream`] and
-/// prevents users from reading and writing to [`DmaStream`] unexpectedly.
-#[derive(Debug, PartialEq, Clone, Copy)]
-pub enum DmaDirection {
-    /// Data flows to the device
-    ToDevice,
-    /// Data flows from the device
-    FromDevice,
-    /// Data flows both from and to the device
-    Bidirectional,
+#[derive(Debug)]
+enum StreamInner {
+    Segment(USegment),
+    Both(KVirtArea, Paddr, USegment),
 }
 
-impl DmaStream {
+/// Disallow users from accessing it by typed memory.
+///
+/// Although, the data inside is untyped and can be accessed by [`VmReader`]
+/// and [`VmWriter`].
+#[cfg(all(target_arch = "x86_64", feature = "cvm_guest"))]
+struct CvmPrivateMeta;
+#[cfg(all(target_arch = "x86_64", feature = "cvm_guest"))]
+impl_frame_meta_for!(CvmPrivateMeta);
```

Comment:
I think this should be:
> Metadata for frames that behind a [`KVirtArea`] of different page table flags.
>
> DMA frames that are shared (in a CVM guest) or that are uncachable (for non-DMA-coherent devices) cannot be accessed via linear mappings. Doing so can cause unexpected side effects. Therefore, we mark these frames as typed memory, meaning users outside of OSTD can no longer safely create [`VmReaders`] or [`VmWriters`] on them.

So I think this metadata should be used for uncached KVA and `DmaCoherent` as well.